File: cpp/234. Palindrome Linked List.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;


struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
// Time: O(n), Space: O(1)
public:
    auto split(ListNode* head) {
        ListNode *fast = head, *slow = head, *pre = nullptr;
        while(fast && fast->next) {
            pre = slow;
            slow = slow->next;
            fast = fast->next->next;
        }
        pre->next = nullptr;
        return pair<ListNode*, ListNode*>(head, slow);
    }
    auto reverse(ListNode* head) {
        ListNode *pre = nullptr, *cur = head, *nxt;
        while(cur) {
            nxt = cur->next;
            cur->next = pre;
            pre = cur;
            cur = nxt;
        }
        return pre;
    }
    bool isPalindrome(ListNode* head) {
        if(!head || !head->next) return true;      
        auto [first, second] = split(head);
        second = reverse(second);

        while(first) {
            if(first->val != second->val) return false;
            first = first->next;
            second = second->next;
        }
        return true;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    
}

File: python/234-Palindrome Linked List.py
from linklist_helper import *

class Solution:
    # def isPalindrome(self, head) -> bool:
    #     "Time: O(n), Space: O(n)"
    #     arr = []
    #     while head:
    #         arr.append(head.val)
    #         head = head.next
    #     return arr == arr[::-1]
    def isPalindrome(self, head) -> bool:
        "Time: O(n), Space: O(1)"
        if head is None or head.next is None: return True
        def find_mid(head):
            slow, fast = head, head
            while fast is not None and fast.next is not None:
                slow = slow.next
                fast = fast.next.next
            if fast is not None: # odd number of nodes, we need the next
                slow = slow.next
            return slow
        def reverse(head):
            cur = head
            prev = None
            while True:
                nxt = cur.next
                cur.next = prev
                if nxt is None:
                    return cur
                prev = cur
                cur = nxt

        mid = find_mid(head)
        right = reverse(mid)
        left = head
        while right is not None:
            if left.val != right.val: return False
            left = left.next
            right = right.next
        return True

head = gen_linklist([1, 2])
result = Solution().isPalindrome(head)
assert result == False

head = gen_linklist([1, 2, 2, 1])
result = Solution().isPalindrome(head)
assert result == True

head = gen_linklist([1])
result = Solution().isPalindrome(head)
assert result == True